lib/SIL/IR/TypeLowering.cpp

```diff
@@ -112,15 +112,25 @@ CaptureKind TypeConverter::getDeclCaptureKind(CapturedValue capture,
   assert(var->hasStorage() &&
          "should not have attempted to directly capture this variable");
 
+  auto &lowering = getTypeLowering(
+      var->getType(), TypeExpansionContext::noOpaqueTypeArchetypesSubstitution(
+                          expansion.getResilienceExpansion()));
+
+  // If this is a noncopyable 'let' constant that is not a shared paramdecl,
+  // then we know it is boxed and want to pass it in its boxed form so we can
+  // obey Swift's capture reference semantics.
+  if (!var->supportsMutation() && lowering.getLoweredType().isPureMoveOnly()) {
```

Comment:
You should also check `!capture.isNoEscape()` here so we don't box captures from nonescaping closures.

Reply:
I am trying that now